flink-libraries/flink-table/src/main/scala/org/apache/flink/api/scala/table/expressionDsl.scala

```diff
@@ -130,19 +130,17 @@ trait ImplicitExpressionOperations {
     * @param endIndex last character of the substring (starting at 1, inclusive)
     * @return substring
     */
-  def substring(beginIndex: Expression, endIndex: Expression) = {
-    Call(BuiltInFunctionNames.SUBSTRING, expr, beginIndex, endIndex)
-  }
+  def substring(beginIndex: Expression, endIndex: Expression) =
+    SubString(expr, beginIndex, endIndex)
 
   /**
     * Creates a substring of the given string beginning at the given index to the end.
     *
     * @param beginIndex first character of the substring (starting at 1, inclusive)
     * @return substring
     */
-  def substring(beginIndex: Expression) = {
-    Call(BuiltInFunctionNames.SUBSTRING, expr, beginIndex)
-  }
+  def substring(beginIndex: Expression) =
+    new SubString(expr, beginIndex)
```

Comment:
Since `SubString` is defined as `case class SubString(str: Expression, begin: Expression, end: Expression)` and so does the generated `apply` method, the `new` here cannot be removed, otherwise complaining of `cannot resolve subString with such signature`